merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/AutoValueMappers.java

```diff
@@ -0,0 +1,166 @@
+      }
+
+      builder.addMethod(
+          MethodSpec
+              .methodBuilder(methodName)
+              .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+              .addTypeVariables(((TypeElement) record).getTypeParameters().stream().map(TypeVariableName::get).toList())
+              .returns(ParameterizedTypeName.get(ClassName.get(ValueMapper.class), TypeName.get(record.asType())))
+              .addParameters(
+                  necessaryMappers
+                      .entrySet()
+                      .stream()
+                      .map(mapperRequest -> ParameterSpec
+                          .builder(
+                              ParameterizedTypeName.get(
+                                  ClassName.get(ValueMapper.class),
+                                  ClassName.get(mapperRequest.getKey()).box()),
+                              mapperRequest.getValue(),
+                              Modifier.FINAL)
+                          .build())
+                      .toList())
+              .addCode(
+                  CodeBlock
+                      .builder()
+                      .add("return new $T<>($>\n", ClassName.get(RecordValueMapper.class))
+                      .add("$L$T.class,\n",
+                           castIfGeneric(record),
+                           ClassName.get((TypeElement) record))
+                      .add("$T.of($>\n", List.class)
+                      .add(CodeBlock.join(
+                          componentToMapperName
+                              .stream()
+                              .map(recordComponent ->
+                                       CodeBlock
+                                           .builder()
+                                           .add(
+                                               "new $T<>($>\n" + "$S,\n" + "$T::$L,\n" + "$L" + "$<)",
+                                               RecordValueMapper.Component.class,
+                                               recordComponent.componentName(),
+                                               ClassName.get((TypeElement) record),
+                                               recordComponent.componentName(),
+                                               recordComponent.mapperName())
+                                           .build())
+                              .toList(),
+                          ",\n"))
+                      .add("$<)$<);")
+                      .build())
+              .build());
+    }
+
+    return JavaFile
+        .builder(typeName.packageName(), builder.build())
+        .skipJavaLangImports(true)
+        .build();
+  }
+
+  private static CodeBlock castIfGeneric(final Element record) {
+    final var typeParameters = ((Parameterizable) record).getTypeParameters();
+    if (typeParameters.isEmpty()) {
+      return CodeBlock.of("");
+    } else {
+      return CodeBlock.of("(Class<$T>) (Object) ", TypeName.get(record.asType()));
```

Comment:
That looks sus. I don't know the use cases for double casts in Java, but in Typescript its usually bad news. Why is `(Object)` necessary?

Reply:
This method is used in this context:

```java
.add("$L$T.class,\n",
    castIfGeneric(record),
    ClassName.get((TypeElement) record))
```

So we can expect the expression at issue to be `(Class<T>) (Object) T.class`, where `T` is some generic type. For instance, `(Class<List<Integer>>) (Object) List.class`.

The expression `List.class` alone has type `Class<LIst>`, not `Class<List<Integer>>`, and attempting to cast directly to `(Class<List<Integer>>)` throws a compile-time error. Via `jshell`:

```
jshell> (List<Integer>) List.class
|  Error:
|  incompatible types: java.lang.Class<java.util.List> cannot be converted to java.util.List<java.lang.Integer>
|  (List<Integer>) List.class
|                  ^--------^
```

Hence the two-phase cast.

Reply:
Seems like if it would give a compile error, but you obfuscate the type enough that it can't, it should just give a runtime cast error.

Reply:
It doesn't give a runtime error though :) This is perfectly safe, because of what `Class` represents. Because generics were added to the language late, there's no way to tell `List.class` what its type parameters are directly. This is "just" letting us get a Class instance that is restricted to use over a particular specialization of type parameters.

Reply:
I stole this idea from here:
https://github.com/NASA-AMMOS/aerie/blob/26a7d5d5f426495bde8fe89a4b900fb1933d38d7/merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/Resolver.java#L71-L73

The goal is to unify the record `.class` with the type of the "projection" functions in the components. The contract is that the RecordValueMapper takes a `Class<R>` (whose main purpose is to provide a constructor via reflection), as well as a list of components, each of which defines a "projection" `Function<R, T>`, where `T` is the type of that component.

https://github.com/NASA-AMMOS/aerie/blob/26a7d5d5f426495bde8fe89a4b900fb1933d38d7/contrib/src/main/java/gov/nasa/jpl/aerie/contrib/serialization/mappers/RecordValueMapper.java#L22-L25

With no casts at all, we get this fun-to-decipher error that says that `R` cannot be simultaneously a `RecordParameter` (the type of the `.class`), and a `RecordParameter<T>` (the input type of the projection function)

```
incompatible types: cannot infer type arguments for RecordValueMapper<>
    return new RecordValueMapper<>(
                                ^
    reason: inference variable R#1 has incompatible equality constraints RecordParameter<T>,R#2,RecordParameter
  where R#1,T,R#2 are type-variables:
    R#1 extends Record declared in class RecordValueMapper
    T extends Object declared in method <T>gov_nasa_jpl_aerie_banananation_activities_ParameterTestActivity_RecordParameter(ValueMapper<T>)
    R#2 extends Object declared in record Component
```

Sounds like this pattern requires a code comment - since it does elicit a perfectly reasonable "what the..." from the reader. I'll see if I can wordsmith something together

Reply:
> Seems like if it would give a compile error, but you obfuscate the type enough that it can't check, it should just give a runtime cast error.

In this particular case - there can be no runtime error because there is no observable difference at runtime between a `Class<MyRecord<T>>` and a `Class<MyRecord>` - both identically erase to `MyRecord.class`. This cast is merely to help us with some compile time bookkeeping: it makes sure that we use this particular `MyRecord.class` to represent `MyRecord<T>` and not simply `MyRecord`.

This does seem to be an artifact of Java's generics being added late, and trying to interoperate with non-generic types - it allows you to cast at the top level:
```java
(ArrayList<List<Integer>>) new ArrayList();
```
But not a level deeper:
```java
(ArrayList<List<Integer>>) new ArrayList<List>();
// Inconvertible types; cannot cast 'java.util.ArrayList<java.util.List>' to 'java.util.ArrayList<java.util.List<java.lang.Integer>>'
```
We could hack around this with an `(Object)` cast. We would know that it's safe because we can see syntactically that the arraylist is empty, and that no other references exist to it - we are merely restricting uses of this value to treat it as an `ArrayList<List<Integer>>`.
```java
(ArrayList<List<Integer>>) (Object) new ArrayList<List>();
```
In that case - casting to `(Object)` would be silly, because we have control over the generic on the right, and can simply adjust it to be what we want it to be. In the case of `MyRecord.class` - that comes out as a `Class<MyRecord>`, and we have no opportunity to restrict it to `Class<MyRecord<T>>` without resorting to this ugly cast.

TL;DR - I suspect that Typescript was designed with generics from its inception, and thus has a cleaner story around them. In Java, there are some rough edges we have to work around...

Reply:
I'm adding a `SAFETY` comment:

```
SAFETY: This cast cannot fail - it merely exists to compensate for an inability
to express a MyRecord<T>.class. Instead, we use MyRecord.class, and cast it, via
Object, to a `Class<MyRecord<T>>`. We need to go via Object because Java generics
do not allow casting a Foo<Bar> to a Foo<Bar<Baz>> - since Bar may be a container
that already contains non-Baz objects. In this case, since MyRecord.class is not
a container, the Java type checker is being overly conservative
```

Reply:
I would add, what we really want is the `Class<F<T>>`, but there is literally no way to do so without a cast. And it's safe to do *this* cast because of what you said above.

`List.class` semantically describes *all* lists, but its default type is too restrictive to actually use it on any but raw lists.